### PR TITLE
Fix `_gen_md5_for_arraylike_obj` to use access pandas DF rows via `iloc` 

### LINKED ISSUE #12269: [BUG] mlflow.evaluate fails with KeyError given DataFrame type input and index not containing 1..5 values
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.13.0
- Tracking server: 2.13.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  
    - Distributor ID:	Ubuntu
    - Description:	Ubuntu 22.04.4 LTS
    - Release:	22.04
- **Python version**: Python 3.10.14
- **yarn version, if running the dev UI**: -

### Describe the problem

The `mlflow.evaluate()` method fails when we evaluate using a DataFrame that has a float type index.

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
import mlflow
from sklearn.datasets import load_wine
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler


target_name = "target"
test_size = 0.2
random_state = 42

experiment_name = "bug_example"
model_type = "classifier"

input_data = load_wine(as_frame=True).frame
input_data = input_data.set_index("malic_acid") # Set a float as index that causes the error

X = input_data.copy()
y = X.pop(target_name)

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=test_size,
    random_state=random_state,
    stratify=y
)

# Create ML pipeline
pipeline = make_pipeline(
    StandardScaler(),
    RandomForestClassifier(random_state=random_state, n_jobs=-1)
)

mlflow.set_experiment(experiment_name)

# Train & evaluate
with mlflow.start_run() as run:
    fitted_pipeline = pipeline.fit(X_train, y_train)
    mlflow.sklearn.log_model(fitted_pipeline, "model")
    model_path = mlflow.get_artifact_uri(artifact_path="model")
    # This step fails
    mlflow.evaluate(
        model=model_path,
        data=input_data,
        targets=target_name,
        model_type=model_type,
    )
```

### Stack trace

```
/lib/python3.10/site-packages/mlflow/models/evaluation/base.py:553: FutureWarning: The behavior of obj[i:j] with a float-dtype index is deprecated. In a future version, this will be treated as positional instead of label-based. For label-based slicing, use obj.loc[i:j] instead
  head_rows = data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
Traceback (most recent call last):
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 3791, in get_loc
    return self._engine.get_loc(casted_key)
  File "index.pyx", line 152, in pandas._libs.index.IndexEngine.get_loc
  File "index.pyx", line 176, in pandas._libs.index.IndexEngine.get_loc
  File "index.pyx", line 219, in pandas._libs.index.IndexEngine._get_loc_duplicates
  File "index.pyx", line 227, in pandas._libs.index.IndexEngine._maybe_get_bool_indexer
  File "index.pyx", line 119, in pandas._libs.index._unpack_bool_indexer
KeyError: 5

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./test_example.py", line 44, in <module>
    mlflow.evaluate(
  File "/lib/python3.10/site-packages/mlflow/models/evaluation/base.py", line 2084, in evaluate
    dataset = data.to_evaluation_dataset(dataset_path, feature_names)
  File "/lib/python3.10/site-packages/mlflow/data/pandas_dataset.py", line 155, in to_evaluation_dataset
    return EvaluationDataset(
  File "/lib/python3.10/site-packages/mlflow/models/evaluation/base.py", line 734, in __init__
    _gen_md5_for_arraylike_obj(md5_gen, self._features_data)
  File "/lib/python3.10/site-packages/mlflow/models/evaluation/base.py", line 553, in _gen_md5_for_arraylike_obj
    head_rows = data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
  File "/lib/python3.10/site-packages/pandas/core/frame.py", line 3876, in __getitem__
    return self._getitem_slice(key)
  File "/lib/python3.10/site-packages/pandas/core/generic.py", line 4284, in _getitem_slice
    slobj = self.index._convert_slice_indexer(key, kind="getitem")
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 4238, in _convert_slice_indexer
    return self.slice_indexer(start, stop, step)
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6602, in slice_indexer
    start_slice, end_slice = self.slice_locs(start, end, step=step)
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6825, in slice_locs
    end_slice = self.get_slice_bound(end, "right")
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6744, in get_slice_bound
    raise err
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 6738, in get_slice_bound
    slc = self.get_loc(label)
  File "/lib/python3.10/site-packages/pandas/core/indexes/base.py", line 3798, in get_loc
    raise KeyError(key) from err
KeyError: 5
```

### Other info / logs

Pandas lazy slicing example with float index.
```
import pandas as pd
import numpy as np
float_indices = np.random.uniform(low=0.5, high=13.3, size=(8,))
df = pd.DataFrame(np.random.randn(8, 4),
                  index=float_indices, columns=['A', 'B', 'C', 'D'])
df.iloc[:5]
# df[:5]
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by julcsii**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/julcsii/mlflow/pull/12410?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12410/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12410
```

</p>
</details>

Accessing DF rows by row count is recommended via iloc. This is to prevent KeyError when the df index is of float.

### Related Issues/PRs

Resolve  #12269 

### What changes are proposed in this pull request?

Introduce a check for the type of the data variable in the `_gen_md5_for_arraylike_obj` method. If it is a pandas DataFrame access the rows via the `iloc` method. 

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by julcsii** [comment]
I was not able to execute the tests in the Github Codespaces. I think it would make sense to add another test for the method where this bug slipped in.

Can somone help with this?

**Comment by julcsii** [comment]
As an alternative fix I was thiking maybe the DF could be transformed to numpy arrays before calling the hash function, so we don't need to use iloc inside the function? Thoughts?

File: mlflow/models/evaluation/base.py [hunk 1]
```diff
@@ -576,8 +576,13 @@ def _gen_md5_for_arraylike_obj(md5_gen, data):
...
     else:
-        head_rows = data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
-        tail_rows = data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]
+        if isinstance(data, pd.DataFrame):
+            # Access rows of pandas Df with iloc
+            head_rows = data.iloc[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
+            tail_rows = data.iloc[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]
+        else:
+            head_rows = data[: EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH]
+            tail_rows = data[-EvaluationDataset.NUM_SAMPLE_ROWS_FOR_HASH :]
```
Comment (harupy) [inline mlflow/models/evaluation/base.py:585]: Is this change testable?

Comment (harupy) [reply mlflow/models/evaluation/base.py:585]: Sounds good. Can we also test a dataframe with shifted indices (e.g. `[100, 101, 102]`)?

Comment (julcsii) [reply mlflow/models/evaluation/base.py:585]: Do you normally develop locally on your machine? I had problems executing tests on the Gitlab Codespaces

Comment (julcsii) [reply mlflow/models/evaluation/base.py:585]: With the shifted integers there is no KeyError. The below works

```
import pandas as pd
import numpy as np
indices = np.random.random_integers(low=100, high=130, size=(8,))
df = pd.DataFrame(np.random.randn(8, 4),
                  index=indices, columns=['A', 'B', 'C', 'D'])
df[:5]
```

Comment (julcsii) [reply mlflow/models/evaluation/base.py:585]: I added a test. I was able to finally more or less set up the developer environment. (Although I still had some errors, but at least the evaluation tests I was able to run)

**Comment by julcsii** [comment]
Now I see this code section  was moved, so I will merge the changes from master

**Comment by julcsii** [comment]
> Now I see this code section was moved, so I will merge the changes from master

Done

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by julcsii** [comment]
> LGTM!

Thanks for your support! Is there anything else to do from my side?

**Comment by julcsii** [comment]
> > LGTM!
> 
> Thanks for your support! Is there anything else to do from my side?

Hi @harupy, do I need to do anything else to get this merged?

**Comment by B-Step62** [comment]
@julcsii Sorry for the late response, the PR looks great and it will be released in the next version. Thank you for much for the contribution!!

**Pull request merged**